wasm/wasm.ir/test/org/jetbrains/kotlin/wasm/ir/SpecTestRunner.kt

```diff
@@ -0,0 +1,262 @@
+            val line: Int,
+            val filename: String,
+            val text: String,
+            val module_type: String,
+        ) : Command()
+
+        @SerialName("action")
+        @Serializable
+        data class ActionCommand(
+            val line: Int,
+            val action: Action,
+            val expected: List<Value>,
+        ) : Command()
+    }
+
+    @Serializable
+    data class Action(
+        val type: String,
+        val field: String,
+        val args: List<Value> = emptyList(),
+        val module: String? = null
+    )
+
+    @Serializable
+    data class Value(
+        val type: String,
+        val value: String? = null
+    )
+}
+
+private fun runSpecTest(specTest: SpecTest, testDir: File, wastFile: File, wabtOptions: List<String>) {
+    for (command in specTest.commands) {
+        when (command) {
+            is SpecTest.Command.Module -> {
+                val wasmFile = File(testDir, command.filename)
+                testWasmFile(wasmFile, testDir.name)
+            }
+            is SpecTest.Command.Register -> {
+            }
+            is SpecTest.Command.AssertReturn -> {
+            }
+            is SpecTest.Command.AssertTrap -> {
+            }
+            is SpecTest.Command.AssertExhaustion -> {
+            }
+            is SpecTest.Command.AssertMalformed -> {
+            }
+            is SpecTest.Command.AssertInvalid -> {
+            }
+            is SpecTest.Command.AssertUnlinkable -> {
+            }
+            is SpecTest.Command.AssertUninstantiable -> {
+            }
+            is SpecTest.Command.ActionCommand -> {
+            }
+        }
+    }
+}
+
+private fun runJsonTest(jsonFile: File, wastFile: File, wabtOptions: List<String>) {
+    require(jsonFile.isFile && jsonFile.exists())
+    val jsonText = jsonFile.readText()
+    val specTest = Json.decodeFromString(SpecTest.serializer(), jsonText)
+    val wasmDir = jsonFile.parentFile!!
+    println("Running json test ${jsonFile.path} ...")
+    runSpecTest(specTest, wasmDir, wastFile, wabtOptions)
+}
+
+val wasmTestSuitePath: String
+    get() = System.getProperty("wasm.testsuite.path")!!
+
+fun testProposal(
+    name: String,
+    wabtOptions: List<String> = listOf("--enable-all"),
+    ignoreFiles: List<String> = emptyList()
+) {
+
+    runSpecTests(name, "$wasmTestSuitePath/proposals/$name", wabtOptions, ignoreFiles)
+}
+
+
+fun runSpecTests(
+    name: String,
+    wastDirectoryPath: String,
+    wabtOptions: List<String>,
+    ignoreFiles: List<String> = emptyList()
+) {
+    // Clean and prepare output dir for spec tests
+    val specTestsDir = File("build/spec-tests/$name")
+    if (specTestsDir.exists())
+        specTestsDir.deleteRecursively()
+    specTestsDir.mkdirs()
+
+    val testSuiteDir = File(wastDirectoryPath)
+    assert(testSuiteDir.isDirectory) { "${testSuiteDir.absolutePath} is not a directory" }
+    for (file in testSuiteDir.listFiles()!!) {
```

Comment:
here perhaps it could also be in the Kotlin style? how does a language developer not preach the sugar that language gives? :) 

  testSuiteDir.listFiles()?.filterNotNull { it.isFile
                        && it.name.endsWith(".wast")
                        && it.name !in ignoreFiles
            }.forEach {
                println("Creating JSON for ${it.path}")
                Wabt.wast2json(it, (it.absoluteFile.nameWithoutExtension.plus(".json")),
                    *wabtOptions.toTypedArray()
                )
                runJsonTest(jsonFile, file, wabtOptions)
            }

Reply:
I'm well aware of these language constructs. But I don't find them improving readability or maintainability in this case too much. I like naming tmp values and not using `it` in complex lambdas.

I would say, what you describe here is "functional" style, not necessarily Kotlin style. We don't have a guideline to use functional style in the project. And short imperative Kotlin looks fine for me.

Your review would be much more helpful if you would spot a problem with architecture, design or implementation logic rather than suggesting using functional style :)

Reply:
ok) i for beginig post my first pull request)